pyhf/__init__.py

```diff
@@ -171,79 +126,57 @@ def set_poi(self,name):
         assert s.stop-s.start == 1
         self.poi_index = s.start
 
-    def add_modifier(self, name, npars, modifier, suggested_init, suggested_bounds):
-        is_constraint = type(modifier) in [histosys_constraint, normsys_constraint, shapesys_constraint]
-        if name in self.par_map:
-            if type(modifier) == normsys_constraint:
-                log.info('accepting existing normsys')
-                return False
-            if type(modifier) == histosys_constraint:
-                log.info('accepting existing histosys')
-                return False
-            if type(modifier) == type(None):
-                log.info('accepting existing unconstrained factor ')
-                return False
-            raise RuntimeError(
-                'shared systematic not implemented yet (processing {})'.format(name))
-        log.info('adding modifier %s (%s new nuisance parameters)', name, npars)
-
+    def add_or_get_modifier(self, channel, sample, modifier_def):
+        """
+        Add a new modifier if it does not exist and return it
+        or get the existing modifier and return it
+
+        Args:
+            channel: current channel object (e.g. from spec)
+            sample: current sample object (e.g. from spec)
+            modifier_def: current modifier definitions (e.g. from spec)
+
+        Returns:
+            modifier object
+
+        Example:
+            modifier = instance.add_or_get_modifier(channel, sample, modifier_def)
+        """
+        # check if modifier type is implemented
+        if modifier_def['type'] not in modifiers.registry:
+            raise RuntimeError('modifier type not implemented yet (processing {0:s})'.format(modifier_def['type']))
+
+        # get modifier class associated with modifier type
+        modifier_cls = modifiers.registry.get(modifier_def['type'])
+
+        # check if modifier of the given name already exists
+        if modifier_def['name'] in self.par_map:
+            if modifier_cls.is_constrained:
+                log.info('accepting existing {0:s} (type: {1:s})'.format(modifier_def['name'], modifier_cls.__name__))
+            else:
+                log.info('accepting existing unconstrained factor')
+            return self.par_map[modifier_def['name']]['modifier']
+
+        # did not return, so create new modifier and return it
+        modifier = modifier_cls(sample['data'], modifier_def['data'])
+        npars = modifier.n_parameters
+        suggested_init = modifier.suggested_init
+        suggested_bounds = modifier.suggested_bounds
+
```

Comment:
I think through the refactoring this has become superfluous, can just to this directly in the `par_map` dict in line 173-174

Reply:
done